### PR TITLE
feat: introduce SchemaConstraintSolver

### PR DESCRIPTION
**Pull request by mlutze**
The same as the last PR. I will follow up to change everywhere it says "record" to "row", but I will leave the type rules as they are.

I've bumped up the solver iteration limit since some schema cases were struggling.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/typer/SchemaConstraintSolver.scala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{Kind, Name, RigidityEnv, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.typer.ConstraintSolver.ResolutionResult
+import ca.uwaterloo.flix.language.phase.typer.TypeConstraint.{Equality, Provenance}
+import ca.uwaterloo.flix.language.phase.unification.Substitution
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+//
+// This is a copy of the SchemaConstraintSolver. Because the shapes are slightly different, we cannot reuse the same solver.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/SchemaConstraintSolver.scala:27]: Copy of what now? :)

**Comment by magnus-madsen** [comment]
Increasing the iteration limit seems dubious-- but lets see what iteration limit is actually needed for the real solver.

**Comment by magnus-madsen** [comment]
Hmm. You iteration threshold is not high enough... Hmm

**Comment by JonathanStarup** [comment]
I think we should have a flag that runs prints the thresholds needed for each function. that would be a good debugging tool

**Comment by mlutze** [comment]
Yeah, I'll adjust the recursive case to do a little more work, which should reduce costs.

**Comment by magnus-madsen** [comment]
> Yeah, I'll adjust the recursive case to do a little more work, which should reduce costs.

Will you do the same for records or no?

**Comment by mlutze** [comment]
> > Yeah, I'll adjust the recursive case to do a little more work, which should reduce costs.
> 
> Will you do the same for records or no?

Yep, I want to keep them exactly 1:1

**Comment by magnus-madsen** [comment]
Apparently it is still unhappy.

**Pull request merged**